substrate/frame/identity/src/lib.rs

```diff
@@ -959,6 +973,100 @@ pub mod pallet {
 			});
 			Ok(())
 		}
+
+		/// Reap an identity, clearing associated storage items and refunding any deposits. Calls
+		/// the `ReapIdentityHandler`, which can be implemented in the runtime. This function is
+		/// very similar to (a) `clear_identity`, but called on a `target` account instead of self;
+		/// and (b) `kill_identity`, but without imposing a slash.
+		///
+		/// Parameters:
+		/// - `target`: The account for which to reap identity state.
+		/// - `num_subs`: Witness data. The maximum number of sub accounts that `target` has.
+		///
+		/// Origin must be the `ReapOrigin`.
+		#[pallet::call_index(15)]
+		#[pallet::weight(T::WeightInfo::reap_identity(*num_subs))]
+		pub fn reap_identity(
+			origin: OriginFor<T>,
+			target: AccountIdLookupOf<T>,
+			num_subs: u32,
+		) -> DispatchResult {
+			// `ReapOrigin` to be set to `EnsureSigned<AccountId>` (i.e. anyone) on chain where we
+			// want to reap data (i.e. Relay) and `EnsureRoot` on chains where we want "normal"
+			// functionality (i.e. the People Chain).
+			T::ReapOrigin::ensure_origin(origin)?;
+			let who = T::Lookup::lookup(target)?;
+
+			// `take` any storage items keyed by `target`
+			let id = <IdentityOf<T>>::take(&who).ok_or(Error::<T>::NotNamed)?;
+			let fields = id.info.additional.len() as u32;
+			let (subs_deposit, sub_ids) = <SubsOf<T>>::take(&who);
+			// check witness data
+			let actual_subs = sub_ids.len() as u32;
+			ensure!(actual_subs <= num_subs, Error::<T>::TooManySubAccounts);
```

Comment:
Why do we need to do this check? Is there any scenario where this function would be called with `num_subs` not being equal to `T::MaxSubAccounts`? We can make it account for that with weight refunds if necessary because IIUC in the current impl the caller will pay proportional to their `num_subs` of choice regardless of how many subs there actually are (and I doubt callers want to check each identity to see the exact numbers).

Reply:
I favor option 1 over option 2 and the reasons are:
- I don't see a scenario where I'd like to reap an identity but only if the number of subs is less than some arbitrary bound I set. In other words, it looks from the outside like `num_subs` is a protection for the user against excessive weight if they don't want to look into the identity they want to reap, which is not a problem if we do weight refunds.
- It would make sense to have the callers look into the struct off-chain if it was computationally intensive, but it's not the case here since we already have access to the number of subs just by accessing the list. We can do this for the user cheaply, it integrates into the system nicely with the weight refund, and it's a simpler interface IMO.

I'm not going to block on this. Just giving my 2c.